### PR TITLE
Fix llama 13B weights loading

### PR DESCRIPTION
**Pull request by python273**
Fixes exception mentioned in https://github.com/geohot/tinygrad/issues/687#issue-1620209645

now fails at
https://github.com/geohot/tinygrad/blob/cbc5a7222aca61f5378465e6b9e1bb678b4affa5/examples/llama.py#L254-L255

### CODE REVIEW

File: extra/utils.py [hunk 1]
```diff
@@ -86,12 +77,14 @@ def persistent_load(self, pid):
...
   return MyPickle(fb0).load(), key_prelookup
 
-def load_single_weight(t:Tensor, myfile, shape, strides, dtype, mmap_allowed=False):
+def load_single_weight(t:Tensor, myfile, shape, strides, dtype, storage_offset, mmap_allowed=False):
   bytes_size = np.dtype(dtype).itemsize
   if t is None:
     myfile.seek(prod(shape) * bytes_size, 1)
     return
 
+  myfile.seek(storage_offset * bytes_size, 1)
```
Comment (python273) [inline extra/utils.py:86]: tldr it was not handling `storage_offset`

**Comment by geohot** [comment]
Does the 13B model work now? It should work on CPU with 64GB of RAM.

File: extra/utils.py (hunk 1 above)
Comment (geohot) [reply extra/utils.py:86]: Hmm, so this is still isn't totally right. mmap uses a different path that doesn't respect this.

**Comment by geohot** [comment]
Can you add tests that save a bunch of torch models that exercise these code paths and test they are loaded correctly?

**Comment by python273** [comment]
> Does the 13B model work now?

no, weights concatenation code is broken

File: test/extra/test_utils.py [hunk 2]
```diff
@@ -18,5 +18,37 @@ def test_fetch_img(self):
     pimg = Image.open(io.BytesIO(img))
     assert pimg.size == (705, 1024)
 
+  def test_fake_torch_load_zipped(self):
+    import torch
+    import numpy as np
+    import tempfile
+    class LayerWithOffset(torch.nn.Module):
+        def __init__(self):
```
Comment (geohot) [inline test/extra/test_utils.py:26]: spacing

**Comment by python273** [comment]
I like my cats. I really do.

Screenshot 2023-03-15 at 00 06 24

File: extra/utils.py [hunk 3]
```diff
@@ -176,10 +176,11 @@ def fake_torch_load(b0):
     key_real[key_lookup.index(k)] = v[0]
 
   # read in the actual data
-  for storage_type, obj_size, tensor, np_shape, np_strides in key_real:
+  for storage_type, obj_size, tensor, np_shape, np_strides, storage_offset in key_real:
     ll = struct.unpack("Q", fb0.read(8))[0]
     assert ll == obj_size, f"size mismatch {ll} != {obj_size}"
-    load_single_weight(tensor, fb0, np_shape, np_strides, storage_type)
+    assert storage_offset == 0, "not implemented"
+    load_single_weight(tensor, fb0, np_shape, np_strides, storage_type, None)
```
Comment (geohot) [inline extra/utils.py:183]: Why is mmap_allowed=None?

**Comment by geohot** [comment]
🔥

File: extra/utils.py (hunk 3 above)
Comment (python273) [reply extra/utils.py:183]: it's `storage_offset`. The model from the test actually breaks old torch format loading, maybe someone else fixes

It fails even before the assert I added, here:
https://github.com/geohot/tinygrad/blob/e28bd11ff142b67ad311e5f9e40a7d66da20618b/extra/utils.py#L174-L176

File: examples/llama.py [hunk 4]
```diff
@@ -241,21 +241,12 @@ def sample(logits, temperature):
...
-      w0t.lazydata.realized = w0
-      w1t.lazydata.realized = w1
-
-      # terrible hacks. force create the output buffer as float16
-      mv.lazydata.realized = Device._buffers[Device.DEFAULT].empty(mv.shape, dtype=w0.dtype)
-
-      if w0.shape[0] != mv.shape[0]: mv.assign(w0t.cat(w1t, dim=0))
-      elif w0.shape[1] != mv.shape[1]: mv.assign(w0t.cat(w1t, dim=1))
+      if w0.shape[0] != mv.shape[0]: mv.assign(w0.cat(w1, dim=0))
+      elif w0.shape[1] != mv.shape[1]: mv.assign(w0.cat(w1, dim=1))
```
Comment (python273) [inline examples/llama.py:249]: btw you need to put hacks back it, currently using a lot of ram, i'm not sure how to fix them

Comment (dmandry) [reply examples/llama.py:249]: I guess it's easier to just use this [merge-meights.py](https://github.com/randaller/llama-chat/blob/main/merge-weights.py) or port it on tinygrad - that will make merged file with weights and you can load it as 7B without any copies in memory.

**Pull request merged**